### DISCUSSION TITLE
Traces without Code Instrumentation using eBPF

### DISCUSSION THREAD

**Discussion by schultyy** [teletrace/teletrace#92] [Show and tell]
Hi All,

I wanted to share a project with you all I've been working on for the past few weeks. 

In June, Osher had posted [this](https://epsagon.slack.com/archives/C8MGU4SSE/p1656001135131849) link in Slack. This is about auto-instrumenting a Go application without changing source code, using eBPF.

This piqued my curiosity and I started to experiment with eBPF. What I want to share with you is my latest project (I have a few others I can also add here later).

What I find interesting about eBPF capabilities is that it allows us to gather additional data on a running process without having to change source code (+ recompilation).

This Proof of Concept demonstrates, how to gather runtime data from a process using eBPF. This data is complementary to what you would usually collect via standard OTEL instrumentation.

For instance, we're able to see whenever a process accepts a new TCP Connection: 

![Screen Shot 2022-09-07 at 4 55 42 PM](https://user-images.githubusercontent.com/235808/188977091-af7c0d25-325c-4bd9-9cde-9378fe6dd96b.png)

Or, we can monitor the files accessed by a specific process:
![Screen Shot 2022-09-07 at 4 56 25 PM](https://user-images.githubusercontent.com/235808/188977258-7a2a9bf2-aac7-49fc-9479-303ecc681157.png)

Another example: Understanding the latency of a DNS Lookup a process needs to perform: 

![Screen Shot 2022-09-07 at 4 57 11 PM](https://user-images.githubusercontent.com/235808/188977469-4dc49b23-6600-4f24-9ee9-a850adb4ef77.png)

---

## How it works

I am gathering data using both bpftrace and ebpf/python. These processes get spun off and managed by a Go program. This Go program gathers the Console Output and converts it into OTEL Spans. 

The Go Program is generic enough to support several bpf scripts. 

## Value

The value in this is how we can enhance our existing data set with data we can collect without source code instrumentation.
eBPF requires root privileges and depending on instrumentation, it may require additional libraries or features enabled in target application (USDT probes, for instance, not used in the example above).

eBPF is heavily used by Facebook and Google in production today - mostly to analyze and route network traffic. 
There are also CNCF projects that focus on specific aspects here. 

I'm happy to discuss specifics and add more use cases.

**Comment by galbash** [comment]
@arbiv @osherv @nemoshlag some of you were also looking a bit around eBPF right? sounds interesting. When @schultyy showed this to me I remembered a vision Sandro once had that you can do profiling and debugging by turning off and on span creation for specific functions at build time. With this we can actually do it at run time - just add a config file for example of what functions should be trace, that configures the eBPF plugin.
(This also connect well with OpAMP, so you can remotely turn on and off span creation for functions 🤯 )

**Comment by schultyy** [comment]
Quick Update: I transferred the repository with my Code into the Epsagon Org: https://github.com/epsagon/go-supervisor

**Reply by schultyy** [reply]
@galbash I'm not sure if everybody can see it - I don't have enough permissions to see or change that for the repo. Could you check?
